### PR TITLE
Move and extend some pattern warnings post-typer

### PR DESCRIPTION
**Pull request by jxnu-liguobin**
~scala/bug#12304~ scala/bug#12438 

If the input type is sealed and the desired type is non-sealed, we get a warning of failure(in domain). This excludes cases where the intended type is sealed, because when both are sealed, they are incompatible and such a warning already exists.

### CODE REVIEW

**Comment by jxnu-liguobin** [comment]
Hi @SethTisue, Can you see if it is feasible?

**Comment by SethTisue** [comment]
@dwijnand is this something you want to look at, or something it would make sense for you and I to look at together?

**Comment by dwijnand** [comment]
Sure, let's have a look together whether this is worth spending time trying to land.

**Comment by SethTisue** [comment]
Jason [describes](https://github.com/scala/scala/pull/9372#issuecomment-741581714) his approach:

> I defer the fruitless type test check for patterns until the patmat phase and also introduce a stronger analysis that checks that if the scrutinee is sealed, at least one member of its domain could match the pattern type. It's not quite there yet as it should limit the stronger anaysis to cases to "strongly sealed" domains where non of the domain is open for further extension, ie leaves are final classes.

**Comment by SethTisue** [comment]
@jxnu-liguobin I see that this is already passing CI. 

So first, what in your view on the status of the test cases in the PR. Do they seem right to you, do they demonstrate that the bug is fixed?

And then second, if there are still shortcomings here, a plausible next step would be to construct one or more failing tests that demonstrate those shortcomings, in order to then go on and attempt to fix them. Do you understand Jason's "It's not quite there yet..." remark?

File: test/files/neg/patmat-sealed-reachable.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class Test {
+
+  // final class, sealed trait
+  def a(c: Option[Int]) = c match { case _: SealedTrait => ; case _ => }
+
+  // final class, SomeClass
+  def b(c: Option[Int]) = c match { case _: SomeClass => ; case _  => }
+
+  // final class, trait, Can't we check it?
+  def c(c: Option[Int]) = c match { case _: UnsealedTrait => ; case _ => }
```
Comment (dwijnand) [inline test/files/neg/patmat-sealed-reachable.scala:17]: This warned in retronym/scala#109, so what changed in 2.13?

File: test/files/neg/patmat-sealed-reachable.check [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  def f(c: O1) = Map("outer" -> Map("inner" -> new Foo)).collect { case (_: Int, foo: Foo) => "should error" }
+                                                                                      ^
+patmat-sealed-reachable.scala:23: warning: fruitless type test: a value of type Option[Int] cannot also be a SomeClass
+  def b(c: Option[Int]) = c match { case _: SomeClass => ; case _  => }
+                                            ^
+patmat-sealed-reachable.scala:20: warning: fruitless type test: a value of type Option[Int] cannot also be a SealedTrait
+  def a(c: Option[Int]) = c match { case _: SealedTrait => ; case _ => }
+                                            ^
+2 warnings
+4 errors
```
Comment (dwijnand) [inline test/files/neg/patmat-sealed-reachable.check]: None of these are the new "Pattern type is not compatible with any of the sealed children of expected type" warning.

**Review by dwijnand — COMMENTED** [review]
Added some notes and thoughts

**Comment by SethTisue** [comment]
Ah, I understand better now:
* scala/bug#12304 wasn't a bug
* Jason's code aims to fix something else, which I've made a new ticket on: scala/bug#12438

File: test/files/neg/patmat-sealed-reachable.scala (hunk 1 above)
Comment (jxnu-liguobin) [reply test/files/neg/patmat-sealed-reachable.scala:17]: This is also my question, so I was confused and wrote a note. I hope someone can answer this:  it is not clear where the difference between 2.12 and 2.13 comes from. I don't know what I missed. 😅

File: test/files/neg/patmat-sealed-reachable.check (hunk 2 above)
Comment (jxnu-liguobin) [reply test/files/neg/patmat-sealed-reachable.check]: Do you mean that these error levels do not need to be tested and should not exist?

**Comment by jxnu-liguobin** [comment]
> So first, what in your view on the status of the test cases in the PR. Do they seem right to you, do they demonstrate that the bug is fixed?

Yes, I read new issue and @dwijnand comments, and I'm sure we haven't reached the end yet. As mentioned in the new issue#12438, there should be more rigorous analysis to solve it. I will try it later.

File: test/files/neg/patmat-sealed-reachable.check (hunk 2 above)
Comment (dwijnand) [reply test/files/neg/patmat-sealed-reachable.check]: No I mean none of these test cases are hitting the new code.  So it looks like the changes aren't doing anything.

Comment (jxnu-liguobin) [reply test/files/neg/patmat-sealed-reachable.check]: Yes, 👍🏻

**Comment by SethTisue** [comment]
@jxnu-liguobin can you explain where this stands currently? do your recent changes resolve the concerns that were raised earlier, and if so how?

**Comment by jxnu-liguobin** [comment]
> @jxnu-liguobin can you explain where this stands currently? do your recent changes resolve the concerns that were raised earlier, and if so how?

I only handled the Scala/bug#12438 independently, and I have a brief description in it. That is, for `Eq(varr, tc: TypeConst)` only when `varr` is sealed.   Becajuse  If `varr` is final, a known error will be raised, and while `tc` is class or sealed trait a known warning will be raised. So, only add patch when `varr` is sealed and `tc` is un sealed trait.

It should be noted that even if `varr` is sealed, subclasses may exist in the itself domain.

**Comment by SethTisue** [comment]
@martijnhoekstra I saw you commented on the ticket. Do you feel able to review these changes? (This isn't code I'm familiar with, and Dale is busy with other things.)

**Comment by SethTisue** [comment]
@lrytz perhaps you could step in here? this is over my head...

**Comment by lrytz** [comment]
@dwijnand are you familiar with this area of match analysis, and could you review this?

**Comment by SethTisue** [comment]
paging @dwijnand — would you like to look at this together during one of our pairing sessions?

**Comment by dwijnand** [comment]
Subsumed by #10657.

**Pull request closed without merging**